docs/pages/guides/using-eslint.mdx

```diff
@@ -49,6 +49,10 @@ module.exports = {
     'universe/native',
     /* @end */
   ],
+  rules: {
+    // Re-enable an important rule disabled by that preset
```

Comment:
It would be good to be precise about the benefits - e.g.
```suggestion
    // Help verify Hooks run when components are re-rendered and the values of variables accessed within Hooks have changed
```

Reply:
It doesn't just help — I'd say that React's behavior without this rule is essentially undefined behavior.

Reply:
I think the proper fix is to have a version that's used internally by Expo with this rule off, and to make the public config have the rule on.

FWIW I'd be surprised if Expo's codebase doesn't have bugs caused by this.

Reply:
It's possible to write valid, fully correct React without lint rules. Happy to merge the actual code change, I'd just prefer for the wording to focus on the benefits and purpose so people understand why the rule exists than to only say it's important.

Reply:
Cases like this are most tricky: https://github.com/facebook/react/issues/27333#issuecomment-1706669249

You don't really think of it as "I need the value of `onClose` to be up-to-date" inside the `IncrementModal` component. There, you're just calling `onClose` so it seems like `[]` is legit. It's not obvious its identity matters somehow.

Then later *someone else* adds some state to `onClose` that's being passed down. Or even further up the chain. They can't know that someone else below has wrong dependencies. And now that state gets "stuck".

The problem is that the person who chose to suppress dependencies and the person who tries to read state might not even know about each other. Or you might not realize those pieces are connected when they're separated by a few layers of indirection. This makes these bugs nightmarish.

And in this case, there is no action of disabling the rule at all — *the user didn't know the rule exists* because the Expo preset disables it despite it being on in the RN preset. (Is that what's happening? If not, my apologies! That's what I inferred.)

Reply:
>It's possible to write valid, fully correct React without lint rules. Happy to merge the actual code change, I'd just prefer for the wording to focus on the benefits and purpose so people understand why the rule exists than to only say it's important.

Conceptually the rule is a part of React; it is not meant as a style nudge. The rule exists because if it's not being followed, React's behavior is not defined. I can't think of a better description than "Follow it because otherwise you'll have gnarly bugs unless you carefully audit every change to the parent component tree". Is there a better way to phrase it?

Reply:
> I think the proper fix is to have a version that's used internally by Expo with this rule off, and to make the public config have the rule on.

Agreed. We are actively working on a more aptly named eslint-config-expo. We will use your feedback for configuring the Hooks linter rules.

> And in this case, there is no action of disabling the rule at all — the user didn't know the rule exists because the Expo preset disables it despite it being on in the RN preset.

eslint-config-universe was historically written for internal use across many different types of projects (servers, websites, apps) and each project overlays its own overrides in its own ESLint config file. It does not use any RN preset. I'm not sure if such a preset even exists (it might now but I don't know if there was one at the time even).

Somewhere along the way developers started externally using eslint-config-universe and it got added to the docs because it does handle a number of cases that ESLint doesn't take care of, like .`{android,ios}.{js,ts}` extensions, for instance.

In any case, eslint-config-expo will be a publicly recommended config that is intended to be composed with a developers own choice of rules (like eslint:recommended or prettier). But if there is a rule that is objectively always useful to have, especially in the context of a new ESLint config and with React Compiler someday(tm), we can include it => which is aligned with your earlier suggestion for there to be a public config with the Hook deps rule on.

In the meantime I have no objections to adding the hooks rule to the docs example, either with an edit like the inline suggestion I posted earlier or no comment.

Reply:
I want to iterate on the comment more — I don't think the important part is re-running Hooks. E.g. when you pass `onClose` down and the state is inside `onClose`, you don't really think in terms of "I need to re-run a hook when state inside changes". So this description makes it sound like this check wouldn't apply in this case.

Reply:
The way I think about the `onClose` example is that the onClose function changes across re-renders. It's a new function in a new lexical environment and therefore a new closure. IIUC the trickiness comes from the fact each re-render of the parent component creates a new lexical environment and the old `onClose` function doesn't close over the new re-render's state.

For comment text, here's another stab: Verify Hooks re-run when their dependencies change. Proactively defends against cases where values accessed within Hooks might be mistakenly thought not to change even though they do, such as callbacks passed in as props that close over the parent component's own props or state. https://react.dev/learn/editor-setup#linting

Or just linking to the React docs is fine too IMO. Then there's more of a canonical place to read about what's going on.

Reply:
>The way I think about the onClose example is that the onClose function changes across re-renders. It's a new function in a new lexical environment and therefore a new closure. 

Yep, this is correct but is largely not understood, even among relatively experienced developers.